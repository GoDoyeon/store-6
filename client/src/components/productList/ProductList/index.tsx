import { FC, useEffect, useState, useRef, useContext } from 'react';

import CategoryFilter from '~/components/productList/CategoryFilter';
import CategoryIdentifier from '~/components/productList/CategoryIdentifier';
import OrderFilter from '~/components/productList/OrderFilter';
import ProductItemContainer from '~/components/productList/ProductItemContainer';
import ScrollToTop from '~/components/productList/ScrollToTop';
import SearchBox from '~/components/productList/SearchBox';

import * as productsAPI from '~/lib/api/products';
import { ProductsGetResponseBody } from '~/lib/api/types';
import FilterContext from '~/lib/contexts/filterContext';
import FetchContext from '~/lib/contexts/fetchContext';
import useIntersection from '~/lib/hooks/useIntersection';
import useScrollPoint from '~/lib/hooks/useScrollPoint';

import fetchModule, {
  finishFetch,
  initFetch,
  INIT_FETCH,
  START_FETCH,
} from '~/stores/fetchModule';
import { setLastPage, setNextPage } from '~/stores/productFilterModule';

import S from './index.style';

// Interface
export interface ProductData {
  idx: number;
  title: string;
  thumbnail: string;
  originPrice: number;
  discountedPrice: number;
}

// Component
const ProductList: FC = () => {
  const TARGET_POINT = 500;
  const DEFAULT_PRODUCTS_AMOUNT = 9;

  const [products, setProducts] = useState<ProductsGetResponseBody[]>([]);
  const listFooterRef = useRef<HTMLDivElement>();

  const { state: filterState, dispatch: productFilterDispatch } =
    useContext(FilterContext);
  const { state: fetchState, dispatch: fetchDispatch } = fetchModule();
  const entry = useIntersection(listFooterRef, {});

  const isScrollPoint = useScrollPoint(TARGET_POINT);

  const fetchProducts = async () => {
    if (filterState.isLastPage) return;

    try {
      const { data } = await productsAPI.getProducts(filterState);

      if (data.length < DEFAULT_PRODUCTS_AMOUNT)
        productFilterDispatch(setLastPage());

      const isNextPageRequest = filterState.page !== 1;
      if (!isNextPageRequest) setProducts(data);
      else setProducts((prev) => [...prev, ...data]);

      setTimeout(
        () => fetchDispatch(finishFetch()),
        fetchState.forcedDelayTime,
      );
    } catch (error) {
      throw new Error(error.data.message);
    }
  };

  useEffect(() => {
    if (fetchState.action === INIT_FETCH) fetchProducts();
    if (fetchState.action === START_FETCH) {
      setTimeout(() => fetchProducts(), fetchState.forcedDelayTime);
    }
  }, [filterState, fetchState.action]);

  useEffect(() => {
    if (fetchState.action !== INIT_FETCH && entry?.isIntersecting) {
      productFilterDispatch(setNextPage());
      fetchDispatch(initFetch());
    }
  }, [entry]);

  return (
    <FetchContext.Provider
      value={{ state: fetchState, dispatch: fetchDispatch }}
    >
      <S.ProductListWrapper>
        <S.LeftSection>
          <CategoryFilter />
          <OrderFilter />
        </S.LeftSection>
        <S.VerticalDivider />
        <S.RightSection>
          <CategoryIdentifier />
          <SearchBox />
          <ProductItemContainer products={products} ref={listFooterRef} />
        </S.RightSection>
        <ScrollToTop isVisible={isScrollPoint} />
      </S.ProductListWrapper>
    </FetchContext.Provider>
  );
};

export default ProductList;
