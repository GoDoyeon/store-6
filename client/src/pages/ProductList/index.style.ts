import styled from 'styled-components';
import { VertLineSVG } from '~/assets';

export const ProductListWrapper = styled.div`
  display: flex;
  width: 100%;
`;

export const LeftSection = styled.section`
  height: 700px;
  position: sticky;
  margin-top: 50px;
  top: 83px;
  display: flex;
  flex-direction: column;
  gap: 100px;
`;

export const RightSection = styled.section`
  width: 100%;
`;

export const VerticalDivider = styled.div`
  width: 1px;
  margin: 0 45px;
  min-height: 1000px;
  background: url(${VertLineSVG}) bottom left no-repeat;
  background-size: cover;
`;
