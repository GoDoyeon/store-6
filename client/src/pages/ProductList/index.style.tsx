import styled from 'styled-components';
import { VertLineSVG } from '~/assets';

export const ProductListWrapper = styled.div`
  display: flex;
  width: 100%;
`;

export const LeftSection = styled.section`
  height: 700px;
  position: sticky;
  margin-top: 50px;
  top: 50px;
`;

export const RightSection = styled.section`
  width: 700px;
`;

export const VerticalDivider = styled.div`
  width: 1px;
  margin: 0 45px;
  min-height: 1000px;
  background: url(${VertLineSVG}) bottom left no-repeat;
  background-size: cover;
`;


