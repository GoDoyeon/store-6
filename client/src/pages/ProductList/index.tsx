import React, { FC } from 'react';

import {
  ProductListWrapper,
  LeftSection,
  RightSection,
  VerticalDivider,
} from './index.style';

const ProductList: FC = () => {
  return (
    <ProductListWrapper>
      <LeftSection>
        <div>카테고리 컴포넌트</div>
        <div>순서 컴포넌트</div>
      </LeftSection>
      <VerticalDivider />
      <RightSection>
        <div>카테고리 현재 선택</div>
        <div>검색</div>
        <div>상품리스트</div>
        <div>스크롤</div>
      </RightSection>
    </ProductListWrapper>
  );
};

export default ProductList;
