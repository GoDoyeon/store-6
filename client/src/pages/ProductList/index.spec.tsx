import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';

import ProductList from './index';

describe('<ProductList />', () => {
  it('should render same with snapshot', () => {
    const { container } = render(<ProductList />);
    expect(container).toMatchSnapshot();
  });
});
