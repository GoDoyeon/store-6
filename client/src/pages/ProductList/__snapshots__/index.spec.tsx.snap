// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductList /> should render same with snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  height: 100%;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  width: 218px;
  height: 100%;
}

.c3 {
  width: 700px;
  height: 100%;
}

.c2 {
  width: 1px;
  height: 100%;
  margin: 0 45px;
  background: url(https:/store-6-bucket.s3.ap-northeast-2.amazonaws.com/common/vertical-line.svg) bottom left no-repeat;
  background-size: cover;
}

<div>
  <div
    class="c0"
  >
    <section
      class="c1"
    >
      <div>
        <button
          type="button"
        >
          sdfsdfsdf
        </button>
      </div>
      <div>
        <span>
          순서 컴포넌트
        </span>
      </div>
    </section>
    <div
      class="c2"
    />
    <section
      class="c3"
    >
      <button
        type="button"
      >
        카테고리 현재 선택
      </button>
      <div>
        검색
      </div>
      <div>
        상품리스트
      </div>
      <div>
        스크롤
      </div>
    </section>
  </div>
</div>
`;
